Experimental: Complex reflection groups (#3342)
oscar-system/Oscar.jl: experimental/ComplexReflectionGroups/docs/src/complex_reflection_groups.md

leaves $(\cdot,\cdot)$ **invariant**, or that $g$ is a **unitrary** transformation with
respect to $(\cdot,\cdot)$, if $(gv,gw) = (v,w)$ for all $v,w \in V$. It is a standard fact
that if $G \subset \mathrm{GL}(V)$ is a *finite* subgroup, then there exists a inner product
on $V$ which is $G$-invariant, i.e. $g$-invariant for all $g \in G$.

Now, let $g \in \mathrm{GL}(V)$ be a reflection. Since $g$ is a assumed to be of finite
order, the group $\langle g \rangle \subset \mathrm{GL}(V)$ is of finite order and so there
exists a $g$-invariant inner product on $V$. We then say that $g$ is a **unitary
reflection** with respect to such an inner product. The orthogonal decomposition

$$V = \mathrm{Ker}(1-g) \oplus \mathrm{Ker}(1-g)^\perp$$

is $g$-stable and

$$\mathrm{Ker}(1-g)^\perp = \mathrm{Im}(1-g)$$

so that we have a $g$-stable orthogonal decomposition

$$V = \mathrm{Ker}(1-g) \oplus \mathrm{Im}(1-g)^\perp \;.$$

Since $\mathrm{Ker}(1-g)$ is a hyperplane, the complement $\mathrm{Im}(1-g)$ is a line. A
non-zero element $\alpha$ on this line is called a **root** of $g$. A root is unique up to
multiplication with a non-zero scalar. We then have $g \alpha = \zeta \alpha$ for some

> primitive $m$-throot of unity $\zeta$, where $m$ is the order of $g$.

Reviewer: ```suggestion
primitive $m$-th root of unity $\zeta$, where $m$ is the order of $g$.
```

!!! warning 
    In OSCAR—like in most computer algebra systems—matrices $A$ act from the *right*
    on vectors $x$ and consequently the kernel of $A$ consists of the vectors $x$ with $xA =
    0$. The same applies to the image and any other construction involving the action. In
    the written literature, however, matrices usually act from the *left*. Hence, to
    transfer results between the computer and the literature, one usually needs to
    *transpose* the matrices.

 

## Showcase
As a demonstration, we create the exceptional complex reflection group $G_{4}$ in the
notation of Shepard & Todd [ST54](@cite). We choose as *model*—meaning a concrete matrix
group—the unitary model from [LT09](@cite). The base field of this model is
$\mathbb{Q}(i,\omega)$ where $i$ is the imaginary unity and $\omega = \exp(2\pi i/3)$. The
group is generated by two unitary reflections. We display the first generator, verify that
it is indeed a unitary reflection, and then we compute additional data like a root, the
hyperplane, and the non-trivial eigenvalue of this reflection. We have implemented an own
type encapsulating such data of a reflection.

> ```@juliarepl

Reviewer 2: ```suggestion
```jldoctest
```
all of these should be doctests (unless they need an eternity to run), so that one gets errors once they get out of date

> ```@jldoctest

Reviewer 2: Those should be without the `@` iirc
```suggestion
```jldoctest
```

  over number field with defining polynomial x^2 + 1
    over rational field

julia> w = gens(W)[1]
[(-1//2*i - 1//2)*ω   (-1//2*i - 1//2)*ω]
[(-1//2*i + 1//2)*ω    (1//2*i - 1//2)*ω]

julia> is_complex_reflection(w)
true

julia> is_unitary(w)
true

julia> b,w_data = is_complex_reflection_with_data(w);

julia> w_data
Complex reflection of order 3 with root (1, (i + 1)*ω + i)

julia> hyperplane(w_data)
[1   (-i - 1)*ω - 1]

julia> eigenvalue(w_data)
-ω - 1
```

One can now run all avaialable matrix group functionality in OSCAR on this group. For 
example one can compute its invariant ring:

```@juliarepl
julia> IR = invariant_ring(W)
Invariant ring
  of matrix group of degree 2 over relative number field

julia> fundamental_invariants(IR)
2-element Vector{MPolyDecRingElem{Hecke.RelSimpleNumFieldElem{AbsSimpleNumFieldElem}, AbstractAlgebra.Generic.MPoly{Hecke.RelSimpleNumFieldElem{AbsSimpleNumFieldElem}}}}:
 x[1]^4 + (-4*ω - 2)*x[1]^2*x[2]^2 + x[2]^4
 x[1]^5*x[2] - x[1]*x[2]^5
```

One feature of our implementation is that we explicitly support *different models*. The
reason is that different models may have different properties. For example, note that in the

> above example the first fundamental invariant has a coefficient in which $\omega$ occurrs,

Reviewer: ```suggestion
above example the first fundamental invariant has a coefficient in which $\omega$ occurs,
```

Author: I hate this word!
